pymc_experimental/distributions/continuous.py

```diff
@@ -0,0 +1,218 @@
+        plt.xlabel('x', fontsize=12)
+        plt.ylabel('f(x)', fontsize=12)
+        plt.legend(loc=1)
+        plt.show()
+
+
+    ========  =========================================================================
+    Support   * :math:`x \in [\mu - \sigma/\xi, +\infty]`, when :math:`\xi > 0`
+              * :math:`x \in \mathbb{R}` when :math:`\xi = 0`
+              * :math:`x \in [-\infty, \mu - \sigma/\xi]`, when :math:`\xi < 0`
+    Mean      * :math:`\mu + \sigma(g_1 - 1)/\xi`, when :math:`\xi \neq 0, \xi < 1`
+              * :math:`\mu + \sigma \gamma`, when :math:`\xi = 0`
+              * :math:`\infty`, when :math:`\xi \geq 1`
+                where :math:`\gamma` is the Euler-Mascheroni constant, and
+                :math:`g_k = \Gamma (1-k\xi)`
+    Variance  * :math:`\sigma^2 (g_2 - g_1^2)/\xi^2`, when :math:`\xi \neq 0, \xi < 0.5`
+              * :math:`\frac{\pi^2}{6} \sigma^2`, when :math:`\xi = 0`
+              * :math:`\infty`, when :math:`\xi \geq 0.5`
+    ========  =========================================================================
+
+    Parameters
+    ----------
+    mu: float
+        Location parameter.
+    sigma: float
+        Scale parameter (sigma > 0).
+    xi: float
+        Shape parameter
+    scipy: bool
+        Whether or not to use the Scipy interpretation of the shape parameter
+        (defaults to `False`).
+
+    References
+    ----------
+    .. [Coles2001] Coles, S.G. (2001).
+        An Introduction to the Statistical Modeling of Extreme Values
+        Springer-Verlag, London
+
+    """
+
+    rv_op = gev
+
+    @classmethod
+    def dist(cls, mu=0, sigma=1, xi=0, scipy=False, **kwargs):
+        # If SciPy, use its parametrization, otherwise convert to standard
+        if scipy:
+            xi = -xi
+        mu = at.as_tensor_variable(floatX(mu))
+        sigma = at.as_tensor_variable(floatX(sigma))
+        xi = at.as_tensor_variable(floatX(xi))
+
+        return super().dist([mu, sigma, xi], **kwargs)
+
+    def logp(value, mu, sigma, xi):
+        """
+        Calculate log-probability of Generalized Extreme Value distribution
+        at specified value.
+
+        Parameters
+        ----------
+        value: numeric
+            Value(s) for which log-probability is calculated. If the log probabilities for multiple
+            values are desired the values must be provided in a numpy array or Aesara tensor
+
+        Returns
+        -------
+        TensorVariable
+        """
+        scaled = (value - mu) / sigma
+
+        logp_expression = at.switch(
+            at.isclose(xi, 0),
+            -at.log(sigma) - scaled - at.exp(-scaled),
+            -at.log(sigma)
+            - ((xi + 1) / xi) * at.log1p(xi * scaled)
+            - at.pow(1 + xi * scaled, -1 / xi),
+        )
+        # bnd = mu - sigma/xi
+        return check_parameters(
+            logp_expression,
+            1 + xi * (value - mu) / sigma > 0,
```

Comment:
This constraint depends on mu so it should probably be in a `switch(constraint, logp, -inf)`. Usually value-dependent constraints are treated elemwise while parameter based constraints are treated blockwise with `check_parameters` (if one fails, everything fails), and we raise an error outside of sampling.

```suggestion
            1 + xi * (value - mu) / sigma > 0,
```

Reply:
I've implemented this now, but not precisely sure the purpose: `check_parameters` is value-based, while `switch(constraint...` is tensor-based, right? Anyway, please have a look.

Reply:
The point is that the logp is usually defined for every possible `value` (although it can be zero for some combinations of value parameters, such as a halfnormal evaluated at -1), so it should never "error" out. Logps can however be undefined for some parameter values, such as a halfnormal with negative sigma. We do that distinction in `PyMC` by using a switch for values and check_parameters for the second.

Reply:
You can increase the test tolerange in `check_logp/ check_locdf`, with the kwarg `decimal`. Do note that you may only be checking a subset of parameter combinations when testing locally, so you can temporarily set `n_samples=-1` to make sure all parameter combinations are tested

Reply:
This is all detailed in the slightly outdated developer guide: https://www.pymc.io/projects/docs/en/latest/contributing/developer_guide_implementing_distribution.html#adding-tests-for-the-logp-logcdf-methods

Reply:
About the checks, is `xi` forced to be between -1 and 1? If so it should be added to `check_parameters`. I can't quite read your interactive output. Calling `check_parameters` should never fail, but it is a symbolic expression. If you want to see if the Assert passes for some inputs you need to `.eval()` it as well.

`at.and_` is the elemwise `and`

Reply:
> You can increase the test tolerange in check_logp/ check_locdf, with the kwarg `decimal`.

`decimal` is an absolute check, defaulting to 6 for float64. So it compares numbers like -1.898391e+61 to a precision of 1e-6 which is 67 orders of magnitude and clearly unnecessary. A relative check like the one I mentioned would compare the ratios of the two numbers to a relative accuracy of 1e-6 instead, which makes more sense.

Reply:
@ricardoV94 good news: Using `at.and_` in `check_parameters` did the trick for the restricted (but practical) domain of $\xi \in [-1,1]$. All checks now pass!

I can expand to the full domain of $\xi \in \mathbb{R}$ if the logp check is made relative instead of absolute as mentioned above.

In any case, thank you so much for your help on this. I think it's been 11 months in the pipeline, so hopefully we can get it merged?!

Reply:
>  A relative check like the one I mentioned would compare the ratios of the two numbers to a relative accuracy of 1e-6 instead, which makes more sense.

I have thought about this sometime ago, but then faced issues. logps vary widely (and non-linearly) between -700xs and +x. More annoying, it's not uncommon to have logps around 0, in which a relative comparison or, on the other hand, absolute comparisons with higher threshold become pretty meaningless.

If you have a more flexible solution that works in general that would be great. Certainly there are more principled ways of doing these accuracy checks.

Reply:
@ricardoV94 `npt.assert_allcose` [here](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) allows for both a relative, `rtol` *and* absolute `atol` tolerance to be set. Example:
```python
>>> x = np.pi*np.array([1e2,1e-1,1e-6,1e-10,1e-100,1e-700])
>>> y = (1 + 1e-6)*x
>>> npt.assert_allclose(x,y)  # will assert
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 5 / 6 (83.3%)
Max absolute difference: 0.00031416
Max relative difference: 9.99999e-07
 x: array([3.141593e+002, 3.141593e-001, 3.141593e-006, 3.141593e-010,
       3.141593e-100, 0.000000e+000])
 y: array([3.141596e+002, 3.141596e-001, 3.141596e-006, 3.141596e-010,
       3.141596e-100, 0.000000e+000])
```
but this does not assert:
```python
>>> npt.assert_allclose(x,y,rtol=1e-6,atol=1e-10)
```

Reply:
Do want to try to open a PR on the main repo to change the logic then?

Reply:
Done: https://github.com/pymc-devs/pymc/pull/6159